### PR TITLE
Handle call cancellation gracefully

### LINKED ISSUE #79: SimpleCache setSnapshot fails when removing a cluster
Setting a new snapshot that removes a cluster fails with the following Exception:
```
controlplane_1  | 2018-11-02 19:29:37 ERROR c.m.fury.controlplane.xds.XdsCache - xds cache rebuild failed
controlplane_1  | io.grpc.StatusRuntimeException: CANCELLED: call already cancelled
controlplane_1  |  at io.grpc.Status.asRuntimeException(Status.java:517)
controlplane_1  |  at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onNext(ServerCalls.java:335)
controlplane_1  |  at io.envoyproxy.controlplane.server.DiscoveryServer$7.send(DiscoveryServer.java:270)
controlplane_1  |  at io.envoyproxy.controlplane.server.DiscoveryServer$7.lambda$null$1(DiscoveryServer.java:214)
controlplane_1  |  at io.envoyproxy.controlplane.cache.Watch.respond(Watch.java:77)
controlplane_1  |  at io.envoyproxy.controlplane.cache.SimpleCache.respond(SimpleCache.java:283)
controlplane_1  |  at io.envoyproxy.controlplane.cache.SimpleCache.lambda$setSnapshot$2(SimpleCache.java:209)
controlplane_1  |  at io.envoyproxy.controlplane.cache.CacheStatusInfo.lambda$watchesRemoveIf$0(CacheStatusInfo.java:142)
controlplane_1  |  at java.base/java.util.Collection.removeIf(Collection.java:492)
controlplane_1  |  at io.envoyproxy.controlplane.cache.CacheStatusInfo.watchesRemoveIf(CacheStatusInfo.java:142)
controlplane_1  |  at io.envoyproxy.controlplane.cache.SimpleCache.setSnapshot(SimpleCache.java:200)
controlplane_1  |  at myapp.controlplane.xds.XdsCache.rebuild(XdsCache.java:125)
controlplane_1  |  at myapp.controlplane.xds.XdsCache.lambda$new$1(XdsCache.java:57)
controlplane_1  |  at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
controlplane_1  |  at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
controlplane_1  |  at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
controlplane_1  |  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
controlplane_1  |  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
controlplane_1  |  at java.base/java.lang.Thread.run(Thread.java:844)
```

It would seem that when removing a cluster, envoy cancels the EDS stream asociated with that cluster, but the control plane still tries to synchronously respond to that stream and fails because it has already been cancelled.
I would expect this scenario to be handled better, at least not fail the `setSnapshot` as this prevents further responses to be sent.

### PR DESCRIPTION
**Pull request by sschepens**
Fixes #79

Immediately cancel watches when `responseObserver` is cancelled.
Catch cancellation exceptions when responding watches.
Correctly handle cancellations on error.

### CODE REVIEW

**Comment by sschepens** [comment]
@snowp @joeyb i'm inclined to remove `WatchCancelledException` from `Watch#respond` it seems as though we sould handle cancellations gracefully as they occur in normal situations, and handling them through exceptions doesn't seem right, too much overhead.

What do you think?

**Comment by sschepens** [comment]
I'm also not entirely sure how to test this changes, I don't know how to trigger a cancellation through tests

**Comment by snowp** [comment]
To test I think you want to call `onError(Status.CANCELED)` on the client stream observer. From what I can tell this will send an RST_STREAM to the server which it will treat as a canceled stream

**Comment by sschepens** [comment]
@snowp added a test for this, please review

File: server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java [hunk 1]
```diff
@@ -804,6 +804,68 @@ public void callbackOnError_doesNotLogError_whenCancelled() {
...
+  @Test
+  public void testCallbacksOnCancelled() throws InterruptedException, ClassNotFoundException {
+    final CountDownLatch streamCloseWithErrorLatch = new CountDownLatch(1);
+    final CountDownLatch watchCreated = new CountDownLatch(1);
+    AtomicReference<Watch> watch = new AtomicReference<>();
+
+    MockDiscoveryServerCallbacks callbacks = new MockDiscoveryServerCallbacks() {
+      @Override
+      public void onStreamCloseWithError(long streamId, String typeUrl, Throwable error) {
+        // watch should already be closed by the time we reporte a stream close error
```
Comment (snowp) [inline server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java]: nit: report

File: server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java [hunk 2]
```diff
@@ -804,6 +804,68 @@ public void callbackOnError_doesNotLogError_whenCancelled() {
...
+        streamCloseWithErrorLatch.countDown();
+      }
+    };
+
+    MockConfigWatcher configWatcher = new MockConfigWatcher(false, createResponses()) {
+      @Override
+      public Watch createWatch(boolean ads, DiscoveryRequest request, Set<String> knownResources,
+                               Consumer<Response> responseConsumer) {
+        watchCreated.countDown();
+        System.out.println("watch created");
```
Comment (snowp) [inline server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java]: remove

Comment (sschepens) [reply server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java]: removed

File: server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java (hunk 1 above)
Comment (sschepens) [reply server/src/test/java/io/envoyproxy/controlplane/server/DiscoveryServerTest.java]: :+1:

**Pull request merged**